library/alloc/src/collections/vec_deque/spec_extend.rs

```diff
@@ -17,19 +17,31 @@ where
     default fn spec_extend(&mut self, mut iter: I) {
         // This function should be the moral equivalent of:
         //
-        //      for item in iter {
-        //          self.push_back(item);
-        //      }
-        while let Some(element) = iter.next() {
-            if self.len() == self.capacity() {
-                let (lower, _) = iter.size_hint();
-                self.reserve(lower.saturating_add(1));
+        // for item in iter {
+        //     self.push_back(item);
+        // }
+        loop {
+            let lower_bound = iter.size_hint().0;
+            if lower_bound != 0 {
+                self.reserve(lower_bound);
             }
 
-            let head = self.head;
-            self.head = self.wrap_add(self.head, 1);
+            match iter.next() {
+                Some(val) => self.push_back(val),
+                None => break,
+            }
+
+            let room = self.capacity() - self.len;
             unsafe {
-                self.buffer_write(head, element);
+                // Safety:
+                // The iter is at most `room` items long,
+                // and `room == self.capacity() - self.len`
+                //   => `self.len + room <= self.capacity()`
+                self.write_iter_wrapping(
+                    self.wrap_idx(self.len),
+                    ByRefSized(&mut iter).take(room),
```

Comment:
I'm not convinced you can actually rely on this for an iterator that's not `TrustedLen` -- there might be some way for it to write one of the methods that means it returns more or less than it should even though the take, since `take` just calls things that something *not* `TrustedLen` could have done wrong.

Reply:
`take` always returns less than `n` elements, no matter the iterator. Problems can occur if the iterator overrides `take` and constructs a `Take` adapter with a different limit. At the moment, this cannot occur because `ByRefSized` is controlled by us and does not forward `take`. If it is removed, however, problems could occur; so the most resilient way is probably to call `Take::new` explicitly.

Reply:
`Take::new` is defined in `core` and is only `pub(in core::iter)`. Is there some way of using it from `alloc` without just making it completely public?

Reply:
The `Take::new` function isn't publicly exported, only the `Take` type itself.

Reply:
We might be able to carefuly audit Take and so on. But considering that this is the default impl and there is a separate `TrustedLen` specialization it's fine to have a merely moderately-optimized loop here instead of trying to squeeze out every single cycle we can save.